File: tools/rtti.py
"""MSVC RTTI recovery: class name -> vtable(s).

Type descriptor holds the mangled name at +16. A Complete Object Locator
stores the descriptor as an image-relative offset. A vtable's slot [-1] holds
the absolute VA of its locator, so the vtable starts 8 bytes after that
pointer.
"""
import re, struct, sys
sys.path.insert(0, "tools")
from pe import Image


class Rtti:
    def __init__(self, img: Image):
        self.img = img

    def descriptors(self, pattern=r"\.\?AV"):
        """{mangled: descriptor_rva}"""
        out = {}
        for name, va, size, raw in self.img.sections:
            if name != ".data" and not name.startswith(".rdata"):
                continue
            blob = self.img.data[raw:raw + size]
            for m in re.finditer(rb"\.\?A[VU][\x21-\x7e]{2,200}?@@", blob):
                text = m.group().decode("ascii")
                out.setdefault(text, va + m.start() - 16)
        return out

    def locators(self, descriptor_rva):
        """COL records whose type-descriptor field points at this descriptor."""
        want = struct.pack("<I", descriptor_rva)
        hits = []
        for name, va, size, raw in self.img.sections:
            if not name.startswith(".rdata") and name != ".data":
                continue
            blob = self.img.data[raw:raw + size]
            start = 0
            while True:
                i = blob.find(want, start)
                if i < 0:
                    break
                start = i + 1
                # field 3 of the locator; the record begins 12 bytes earlier
                col = va + i - 12
                sig = self.img.u32(col)
                if sig in (0, 1):
                    hits.append(col)
        return hits

    def vtables(self, col_rva):
        """Vtables whose [-1] slot stores this locator's VA."""
        want = struct.pack("<Q", self.img.base + col_rva)
        hits = []
        for name, va, size, raw in self.img.sections:
            if not name.startswith(".rdata") and name != ".data":
                continue
            blob = self.img.data[raw:raw + size]
            start = 0
            while True:
                i = blob.find(want, start)
                if i < 0:
                    break
                start = i + 1
                hits.append(va + i + 8)
        return hits

    def find(self, class_name):
        """[(mangled, descriptor, [(col, [vtable, ...]), ...]), ...]"""
        out = []
        for mangled, drva in self.descriptors().items():
            if class_name in mangled:
                cols = [(c, self.vtables(c)) for c in self.locators(drva)]
                out.append((mangled, drva, cols))
        return out

    def methods(self, vtable_rva, limit=64):
        """Function pointers from a vtable, stopping at the first non-code."""
        out = []
        for n in range(limit):
            va = self.img.u64(vtable_rva + n * 8)
            if not va or va < self.img.base:
                break
            rva = va - self.img.base
            if self.img.section_of(rva) != ".text":
                break
            out.append(rva)
        return out


if __name__ == "__main__":
    img = Image()
    r = Rtti(img)
    for needle in sys.argv[1:] or ["AiPickUpOrder", "AiPickUpState", "PickUpWeapon", "WeaponSlotScriptInfo"]:
        print(f"===== {needle}")
        for mangled, drva, cols in r.find(needle):
            print(f"  {mangled}  descriptor {drva:#x}")
            for col, vts in cols:
                names = " ".join(f"{v:#x}" for v in vts) or "(none)"
                print(f"    locator {col:#x}  vtable {names}")

File: crates/core/src/rtti.rs
//! MSVC RTTI, read from a live module: class name to vtable to methods.
//!
//! This is `tools/rtti.py`'s walk, on a module image in memory rather than a PE
//! on disk. The layout is the same:
//!
//! - a *type descriptor* holds the mangled name at `+16` (`.?AVWidget@@`);
//! - a *complete object locator* stores the descriptor as an image-relative
//!   offset in its fourth field, and is found by searching for that offset;
//! - a *vtable*'s `[-1]` slot holds the locator's absolute address, so its
//!   methods begin eight bytes after that pointer.
//!
//! What RTTI does not carry is a *name* per method: the slots are in
//! declaration order with no labels. `find` gives the addresses; naming them is
//! the loader's `rtti_method`, from a table the analysis tools produce.

/// One class's vtable, with the methods that could be read from it.
pub struct Vtable {
    /// the mangled name, `.?AVWidget@@`
    pub mangled: String,
    /// the descriptor's rva
    pub descriptor: usize,
    /// the complete object locator's rva
    pub locator: usize,
    /// the first method slot, just past the locator pointer
    pub methods_at: usize,
    /// method rvas, in slot order, stopping at the first non-code slot
    pub methods: Vec<usize>,
}

/// The classes whose mangled name contains `class`, each with its vtable and
/// methods. `is_code` says whether a method rva is executable, which is how the
/// walk knows where a vtable ends and the next data begins.
pub fn find(
    image: &[u8],
    base: usize,
    class: &str,
    is_code: &dyn Fn(usize) -> bool,
) -> Vec<Vtable> {
    let mut out = Vec::new();
    for (mangled, descriptor) in descriptors(image) {
        if !mangled.contains(class) {
            continue;
        }
        for locator in locators(image, descriptor) {
            for methods_at in vtables(image, base, locator) {
                let methods = methods(image, base, methods_at, is_code);
                out.push(Vtable {
                    mangled: mangled.clone(),
                    descriptor,
                    locator,
                    methods_at,
                    methods,
                });
            }
        }
    }
    out
}

/// The method address in `slot` (zero-based) of the first vtable of `class`, as
/// an rva, or `None`.
pub fn method(
    image: &[u8],
    base: usize,
    class: &str,
    slot: usize,
    is_code: &dyn Fn(usize) -> bool,
) -> Option<usize> {
    let found = find(image, base, class, is_code);
    found
        .first()
        .and_then(|vtable| vtable.methods.get(slot).copied())
}

/// Every `.?AV...@@` / `.?AU...@@` string, as (mangled, descriptor rva). The
/// descriptor is sixteen bytes before the name.
fn descriptors(image: &[u8]) -> Vec<(String, usize)> {
    let mut out = Vec::new();
    let mut at = 0usize;
    while let Some(found) = image[at..].windows(3).position(|w| w == b".?A") {
        let start = at + found;
        let kind = image.get(start + 3);
        if kind == Some(&b'V') || kind == Some(&b'U') {
            if let Some(end) = image[start..].windows(2).position(|w| w == b"@@") {
                let text_end = start + end + 2;
                if start >= 16 {
                    if let Ok(text) = std::str::from_utf8(&image[start..text_end]) {
                        out.push((text.to_string(), start - 16));
                    }
                }
                at = text_end;
                continue;
            }
        }
        at = start + 1;
    }
    out
}

/// The complete object locators whose descriptor field points at `descriptor`.
fn locators(image: &[u8], descriptor: usize) -> Vec<usize> {
    let want = (descriptor as u32).to_le_bytes();
    let mut out = Vec::new();
    let mut at = 0usize;
    while let Some(found) = image[at..].windows(4).position(|w| w == want) {
        let field = at + found;
        if field >= 12 {
            let locator = field - 12;
            // the record's first field is a signature of 0 or 1
            if let Some(signature) = u32_at(image, locator) {
                if signature <= 1 {
                    out.push(locator);
                }
            }
        }
        at = field + 1;
    }
    out
}

/// The vtables whose `[-1]` slot holds this locator's absolute address.
fn vtables(image: &[u8], base: usize, locator: usize) -> Vec<usize> {
    let want = (base.wrapping_add(locator) as u64).to_le_bytes();
    let mut out = Vec::new();
    let mut at = 0usize;
    while let Some(found) = image[at..].windows(8).position(|w| w == want) {
        out.push(at + found + 8);
        at = at + found + 1;
    }
    out
}

/// The method rvas from a vtable, stopping at the first slot that is null,
/// outside the module, or not code.
fn methods(
    image: &[u8],
    base: usize,
    mut at: usize,
    is_code: &dyn Fn(usize) -> bool,
) -> Vec<usize> {
    let mut out = Vec::new();
    while let Some(value) = u64_at(image, at) {
        if value < base as u64 {
            break;
        }
        let rva = value as usize - base;
        if rva >= image.len() || !is_code(rva) {
            break;
        }
        out.push(rva);
        at += 8;
    }
    out
}

fn u32_at(image: &[u8], at: usize) -> Option<u32> {
    image
        .get(at..at + 4)
        .map(|b| u32::from_le_bytes(b.try_into().unwrap()))
}

fn u64_at(image: &[u8], at: usize) -> Option<u64> {
    image
        .get(at..at + 8)
        .map(|b| u64::from_le_bytes(b.try_into().unwrap()))
}

#[cfg(test)]
mod tests {
    use super::*;

    /// A stand-in module with one class, `.?AVWidget@@`, its locator and a
    /// three-method vtable. `text` is where the walk treats pointers as code.
    fn stand_in() -> (Vec<u8>, usize) {
        const BASE: usize = 0x1800_0000;
        const TEXT: std::ops::Range<usize> = 0x1000..0x2000;
        let mut image = vec![0u8; 0x3000];
        // a few executable bytes for the method targets
        image[TEXT][..0x40].fill(0xcc);
        let descriptor = 0x2000;
        image[descriptor + 16..descriptor + 16 + 12].copy_from_slice(b".?AVWidget@@");
        let locator = 0x2100;
        image[locator..locator + 4].copy_from_slice(&0u32.to_le_bytes());
        image[locator + 12..locator + 16].copy_from_slice(&(descriptor as u32).to_le_bytes());
        let methods_at = 0x2208;
        let pointer = ((BASE + locator) as u64).to_le_bytes();
        image[0x2200..0x2208].copy_from_slice(&pointer);
        for (n, method) in [0x1000usize, 0x1010, 0x1020].into_iter().enumerate() {
            let pointer = ((BASE + method) as u64).to_le_bytes();
            image[methods_at + n * 8..methods_at + n * 8 + 8].copy_from_slice(&pointer);
        }
        (image, BASE)
    }

    #[test]
    fn finds_a_class_and_its_methods() {
        let (image, base) = stand_in();
        let is_code = |rva: usize| (0x1000..0x2000).contains(&rva);
        let found = find(&image, base, "Widget", &is_code);
        assert_eq!(found.len(), 1);
        assert_eq!(found[0].mangled, ".?AVWidget@@");
        assert_eq!(found[0].locator, 0x2100);
        assert_eq!(found[0].methods, vec![0x1000, 0x1010, 0x1020]);
        assert_eq!(method(&image, base, "Widget", 1, &is_code), Some(0x1010));
        assert_eq!(method(&image, base, "Widget", 3, &is_code), None);
        assert!(find(&image, base, "Nothing", &is_code).is_empty());
    }

    #[test]
    fn a_method_off_the_code_stops_the_walk() {
        let (mut image, base) = stand_in();
        // the middle slot points outside .text, so only the first is a method
        let outside = ((base + 0x2fff) as u64).to_le_bytes();
        image[0x2210..0x2218].copy_from_slice(&outside);
        let is_code = |rva: usize| (0x1000..0x2000).contains(&rva);
        let found = find(&image, base, "Widget", &is_code);
        assert_eq!(found[0].methods, vec![0x1000]);
    }
}
